File: plain-old-scala-factory-design-pattern/src/main/scala/com/thoughtworks/plainoldscalafactorydesignpattern/MonadFactory.scala
package com.thoughtworks.plainoldscalafactorydesignpattern
import language.higherKinds

/**
  * @author 杨博 (Yang Bo)
  */
trait MonadFactory extends FunctorFactory {

  type Facade[+A] <: Monad[A]

  trait Monad[+A] extends Any with Functor[A] {

    def map[B](mapper: (A) => B): Facade[B] = {
      flatMap { a =>
        pure(mapper(a))
      }
    }

    def flatMap[B](mapper: A => Facade[B]): Facade[B]

  }

  def pure[A](a: A): Facade[A]

}

File: plain-old-java-factory-design-pattern/src/main/java/com/thoughtworks/plainoldjavafactorydesignpattern/MonadFactory.java
package com.thoughtworks.plainoldjavafactorydesignpattern;


import java.util.function.Function;

/**
 * @author 杨博 (Yang Bo)
 */
public interface MonadFactory {

    interface Monad<A> {

        MonadFactory getFactory();

        <B> Monad<B> flatMap(Function<A, Monad<B>> mapper);

        default <B> Monad<B> map(Function<A, B> mapper) {
            return flatMap(a -> getFactory().newInstance(mapper.apply(a)));
        }

    }

    <A> Monad<A> newInstance(A a);
}
